### ISSUE TITLE
Always resolvable `Val` alternative with no `Auto` variant

### ISSUE THREAD

**Issue by ickshonpe** [bevyengine/bevy#9688]
## What problem does this solve or what need does it fill?

Val's Auto variant doesn't make sense for certain Style fields, such as the gap properties. In those cases internally `Auto` just gets mapped to Taffy's LengthPercentage::Points(0.)

This might not seem that significant but it's been bothering me intensely the last year. It's never ideal to have multiple representations mapping to the same underlying value like this.

## What solution would you like?

My proposal is PR #8096 which I submitted a while ago. It needs an update to support the recent changes to `Val`, and the macro that implements some of the helper functions could be improved maybe.

Making `Margin`, `Padding` and `Border` into separate types might be controversial. I'm not bothered if we have a generic `UiRect<T>` or `UiRect` (with `Val` edge values) and `NumRect` (with `Num`s) instead. The only important change is adding the always resolvable type.

## What alternative(s) have you considered?

There are some discussions at: https://github.com/bevyengine/bevy/pull/7656, https://github.com/bevyengine/bevy/pull/7969, https://github.com/bevyengine/bevy/pull/7710, https://github.com/bevyengine/bevy/pull/7569.

**Comment by nicoburns** [comment]
Hmm... I'm a little concerned about discoverability here. I wonder if we could add the type you suggest, and then type alias `Margin`, `Padding`, `Border`, etc to the appropriate type?

It may also be worth giving some consideration to how we might want to handle this in a future where there are additional variants for some properties such as width/height (e.g. `MinContent`, `MaxContent`, `FitContentPx`, `FitContentPercent`, and possibly a morphorm-like `Stretch` or `Fr` unit).

**Comment by ickshonpe** [comment]
Yeah I'm rethinking the separate `Margin`, `Padding`, `Border` thing again. The main motivation was that it would allow us to have separate docs for each type. I think probably it's not worth it, and better just to make `UiRect` generic. It's easy to write these things but maintaining them is another matter. I don't want to repeat the recent Val maintenance I did, spend hours fixing lots of trivial but buggy functions and then realize after they've been fixed that they are completely useless anyway.

My personal preference has always been to flatten everything. `margin_left`, `border_thickness_left`, etc fields on `Style`, with helper functions to set values on multiple edges.

**Comment by viridia** [comment]
Just as an aside, in the CSS documentation, this type is called a "Length". I prefer that over "Val", as it more clearly narrows the meaning (a "value" can be almost anything).
